File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RegionPreferenceView.java
package com.garpr.android.views;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.support.annotation.AttrRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StyleRes;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.view.View;

import com.garpr.android.App;
import com.garpr.android.R;
import com.garpr.android.activities.SetRegionActivity;
import com.garpr.android.misc.Heartbeat;
import com.garpr.android.misc.RegionManager;

import javax.inject.Inject;

public class RegionPreferenceView extends SimplePreferenceView implements Heartbeat,
        RegionManager.OnRegionChangeListener, View.OnClickListener {

    @Inject
    RegionManager mRegionManager;


    public RegionPreferenceView(@NonNull final Context context,
            @Nullable final AttributeSet attrs) {
        super(context, attrs);
    }

    public RegionPreferenceView(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public RegionPreferenceView(@NonNull final Context context, @Nullable final AttributeSet attrs,
            @AttrRes final int defStyleAttr, @StyleRes final int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    public boolean isAlive() {
        return ViewCompat.isAttachedToWindow(this);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        if (isInEditMode()) {
            return;
        }

        mRegionManager.addListener(this);
        refresh();
    }

    @Override
    public void onClick(final View v) {
        final Context context = getContext();
        context.startActivity(new Intent(context, SetRegionActivity.class));
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        mRegionManager.removeListener(this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        if (!isInEditMode()) {
            App.get().getAppComponent().inject(this);
        }

        setOnClickListener(this);
        setTitleText(R.string.set_your_region);

        if (isInEditMode()) {
            return;
        }

        mRegionManager.addListener(this);
        refresh();
    }

    @Override
    public void onRegionChange(final RegionManager regionManager) {
        if (isAlive()) {
            refresh();
        }
    }

    @Override
    public void refresh() {
        super.refresh();

        setDescriptionText(mRegionManager.getRegion().getDisplayName());
    }

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/views/RegionPreferenceView.kt
package com.garpr.android.views

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.core.content.ContextCompat
import com.garpr.android.R
import com.garpr.android.activities.SetRegionActivity
import com.garpr.android.extensions.activity
import com.garpr.android.extensions.appComponent
import com.garpr.android.managers.RegionManager
import com.garpr.android.misc.RequestCodes
import javax.inject.Inject

class RegionPreferenceView @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : SimplePreferenceView(context, attrs), RegionManager.OnRegionChangeListener,
        View.OnClickListener {

    @Inject
    protected lateinit var regionManager: RegionManager

    init {
        titleText = resources.getText(R.string.region)

        if (isInEditMode) {
            descriptionText = resources.getString(R.string.region_endpoint_format,
                    resources.getString(R.string.norcal), resources.getString(R.string.gar_pr))
        }

        imageDrawable = ContextCompat.getDrawable(context, R.drawable.ic_location_on_white_24dp)

        setOnClickListener(this)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

        if (isInEditMode) {
            return
        }

        regionManager.addListener(this)
        refresh()
    }

    override fun onClick(v: View) {
        val activity = this.activity

        if (activity == null) {
            context.startActivity(SetRegionActivity.getLaunchIntent(context))
        } else {
            activity.startActivityForResult(SetRegionActivity.getLaunchIntent(activity),
                    RequestCodes.CHANGE_REGION.value)
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        regionManager.removeListener(this)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()

        if (!isInEditMode) {
            appComponent.inject(this)
            regionManager.addListener(this)
            refresh()
        }
    }

    override fun onRegionChange(regionManager: RegionManager) {
        if (isAlive) {
            refresh()
        }
    }

    override fun refresh() {
        super.refresh()

        val region = regionManager.getRegion(context)
        descriptionText = resources.getString(R.string.region_endpoint_format,
                region.displayName, resources.getString(region.endpoint.title))
    }

}
